### PR TITLE
[5.8] Fix issue fake dispatcher (The 3th way)

### PR DESCRIPTION
**Pull request by imanghafoori1**
Look at the https://github.com/laravel/framework/pull/27480 for description.

### CODE REVIEW

File: src/Illuminate/Auth/SessionGuard.php [hunk 1]
```diff
@@ -705,16 +705,16 @@ public function setCookieJar(CookieJar $cookie)
...
 
     /**
      * Set the event dispatcher instance.
      *
-     * @param  \Illuminate\Contracts\Events\Dispatcher  $events
+     * @param  \Closure  $events
      * @return void
      */
-    public function setDispatcher(Dispatcher $events)
+    public function setDispatcher(\Closure $events)
```
Comment (GrahamCampbell) [inline src/Illuminate/Auth/SessionGuard.php]: Maybe use `callable` here?

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

**Comment by imanghafoori1** [comment]
I think some variable names should also change.
For example the argument for `setDispatcher` has no longer a relevant name.

**Comment by driesvints** [comment]
Can we please thoroughly consider if this is the right way? These changes seem unwanted to me. You're swapping out a contract with a very loose implementation.

**Comment by imanghafoori1** [comment]
@driesvints Of course , I have tried to fix the issue in 3 different ways, the 2 others are closed.
The idea here is to defer resolving of dispatcher to the last moment.

Anyway fill free to close it up if you do not like it.

**Comment by imanghafoori1** [comment]
@driesvints How on earth framework users need to call that setDispatcher method. (It's Extremely rare)
It seems to be only called by the internals of the laravel.

**Comment by taylorotwell** [comment]
See comment on your other PR.

**Comment by imanghafoori1** [comment]
@taylorotwell 
- Please reconsider this PR, since we have 3 guards in the core and only the `sessionGuard` dispatches events.
The 2 other guards do not even have a `setDispatcher` method on them.

- This PR covers any guard that can accept a dispatcher through a `setDispatcher` method and is not coupled or know any thing about a specific guard.
The same thing is true for the other PR, https://github.com/laravel/framework/pull/27480

**Pull request closed without merging**